README.md
[![pub package](https://img.shields.io/pub/v/flutter_d2go.svg)](https://pub.dartlang.org/packages/flutter_d2go)

# flutter_d2go

Flutter Plugin inferring using [d2go](https://github.com/facebookresearch/d2go), the mobile model of [detectron2](https://github.com/facebookresearch/detectron2).

## Features

- Get class and boundary box by object detection (Android and iOS)
- Get keypoints by keypoint estimation (Android and iOS)
- Get mask data by instance segmentation (Android only)
- Live inference for camera stream images (Android only)

## Preview

- Live inference for camera stream images

  ![](images/live.gif)
  ![](images/keypoint_live.gif)

- Object detection and instance segmentation

  ![](images/preview.gif)

- Keypoints estimation

  ![](images/keypoints.png)

## Installation

Add flutter_d2go to your `pubspec.yaml`.

Put the d2go model and class file in the assets directory.

```yaml
assets:
  - assets/models/d2go.pt
  - assets/models/classes.txt
```

## Usage

### 1. Load model and classes

The model is in Pytorch format.  
The format of classes file is [here](example/assets/models/classes.txt).

```dart
await FlutterD2go.loadModel(
    modelPath: 'assets/models/d2go.pt',     // required
    labelPath: 'assets/models/classes.txt', // required
);
```

### 2. Get static image predictions

```dart
List<Map<String, dynamic>> output = await FlutterD2go.getImagePrediction(
    image: image,           // required File(dart:io) image
    width: 320,             // defaults to 320
    height: 320,            // defaults to 320
    mean: [0.0, 0.0, 0.0],  // defaults to [0.0, 0.0, 0.0]
    std: [1.0, 1.0, 1.0],   // defaults to [1.0, 1.0, 1.0]
    minScore: 0.7,          // defaults to 0.5
);
```

### 3. Get stream images predictions

```dart
List<Map<String, dynamic>> output = await FlutterD2go.getStreamImagePrediction(
    imageBytesList: cameraImage.planes.map((plane) => plane.bytes).toList(),             // required List<Uint8List> image byte array
    imageBytesPerPixel: cameraImage.planes.map((plane) => plane.bytesPerPixel).toList(), // default to [1, 2, 2]
    width: cameraImage.width,               // default to 720
    height: cameraImage.height,             // default to 1280
    inputWidth: 320,                        // defaults to 320
    inputHeight: 320,                       // defaults to 320
    mean: [0.0, 0.0, 0.0],                  // defaults to [0.0, 0.0, 0.0]
    std: [1.0, 1.0, 1.0],                   // defaults to [1.0, 1.0, 1.0]
    minScore: 0.7,                          // default to 0.5
    rotation: 90,                           // default to 0
);
```

### Predictions `output` format

`rect` is the scale of the original image.  
`mask` and `keypoints` depend on whether the d2go model has mask and keypoints.

`mask` will be a Uint8List of bitmap images bytes.
`keypoints` will be a list of 17 (x, y).

```dart
[
  {
    "rect": {
      "left": 74.65713500976562,
      "top": 76.94147491455078,
      "right": 350.64324951171875,
      "bottom": 323.0279846191406
    },
    "mask": [66, 77, 122, 0, 0, 0, 0, 0, 0, 0, 122, ...],
    "keypoints": [[117.14504, 77.277405], [122.74037, 73.53044], [105.95437, 73.53044], ...],
    "confidenceInClass": 0.985002338886261,
    "detectedClass": "bicycle"
  }, // For each instance
...
]
```

## Issues

Please contact [here](https://github.com/tsubauaaa/flutter_d2go/issues) if you find any bugs or features you would like to add.


CHANGELOG.md
## 0.6.1

- Removed unnecessary import statements.

## 0.6.0

- upgrade Native library to torch 1.10 (ios) / 1.11 (android).

## 0.5.1

- Changes for flutter 3.0 support.

## 0.5.0

- Run keypoints estimates on iOS.

## 0.4.9

- Eliminates path_provider and path library dependencies.

## 0.4.8

- Implemented stream image inference function.

## 0.3.0

- Implemented keypoint estimation function.

## 0.2.1

- The preview image of readme is displayed.

## 0.2.0

- Enabled to handle mask data of instance segmentation.

## 0.0.4

- I am currently trying to implement segmentation. (Segmentation has not been implemented yet.) I tried to separate the Object Detection and Segmentation methods, but decided to use one method instead.

## 0.0.3

- Changed the method name from getImagePrediction to getObjectDetectionResult

## 0.0.1

- initial release


lib/flutter_d2go.dart
import 'dart:async';
import 'dart:io';

import 'package:flutter/services.dart';

/// A list of Y-byte, Cb-byte, and Cr-byte pixel strides.
const List<int> kBytesPerPixel = [1, 2, 2];

/// Camera stream image width size.
const int kWidth = 720;

/// Camera stream image height size.
const int kHeight = 1280;

/// Width size to resize for inference.
const int kInputWidth = 320;

/// Height size to resize for inference.
const int kInputHeight = 320;

/// mean for normalization.
const List<double> kNormMean = [0.0, 0.0, 0.0];

/// Standard deviation for normalization.
const List<double> kNormStd = [1.0, 1.0, 1.0];

/// Threshold of the inference result
const double kMinScore = 0.5;

/// Tilt according to the orientation of the image to be inferred.
const int kRotation = 0;

/// Infer using d2go in flutter.
///
/// Inference can be done for a static image and camera stream images.
///
/// This class has a static method that performs each inference process.
class FlutterD2go {
  static const MethodChannel _channel =
      MethodChannel('tsubauaaa.com/flutter_d2go');

  /// Load d2go model and label file.
  ///
  /// Receive the d2go relative path [modelPath] and [labelPath] in Flutter's asset.
  ///
  /// A method that calls loadModel with invokeMethod and creates pytorch module and
  /// label ArrayList on the Native side.
  ///
  /// Returns `success` string on success and error message on failure.
  static Future<String?> loadModel(
      {required String modelPath, required String labelPath}) async {
    return await _channel.invokeMethod('loadModel', {
      'modelPath': modelPath,
      'labelPath': labelPath,
    });
  }

  /// Get the inference result of a static image.
  ///
  /// Using the image file [image] (required) for inference, the image size for inference [inputWidth], [inputHeight],
  /// the mean [mean] and standard deviation [std] for image normalization,
  /// the threshold of the inference result [minScore], and get the inference result.
  ///
  /// The format is List of
  /// `{ "rect": { "left": double, "top": double, "right": double, "bottom": double }, "mask": Uint8List,
  /// "keypoints": [[double, double], [double, double], [double, double], [double, double], ...],
  /// "confidenceInClass": double, "detectedClass": String }`.
  ///
  /// "mask" and "keypoints" do not exist on some models.
  static Future<List> getImagePrediction({
    required File image,
    int inputWidth = kInputWidth,
    int inputHeight = kInputHeight,
    List<double> mean = kNormMean,
    List<double> std = kNormStd,
    double minScore = kMinScore,
  }) async {
    final List prediction = await _channel.invokeMethod(
      'predictImage',
      {
        'image': image.readAsBytesSync(),
        'inputWidth': inputWidth,
        'inputHeight': inputHeight,
        'mean': mean,
        'std': std,
        'minScore': minScore,
      },
    );

    return prediction;
  }

  /// Get the inference result of the camera stream image.
  ///
  /// Using the camera stream image [imageBytesList] (required), [imageBytesPerPixel] for inference,
  /// the stream image size [width], [height],
  /// the image size for inference [inputWidth], [inputHeight],
  /// the mean [mean] and standard deviation [std] for image normalization,
  /// the threshold of the inference result [minScore],
  /// the tilt according to the orientation of the image to be inferred [rotation],
  /// and get the inference result.
  /// The format is List of { "rect": { "left": double, "top": double, "right": double, "bottom": double },
  ///                         "mask": Uint8List,
  ///                         "keypoints": [[double, double], [double, double], [double, double], [double, double], ...],
  ///                         "confidenceInClass": double, "detectedClass": String }. "mask" and "keypoints" do not exist on some models.
  static Future<List> getStreamImagePrediction({
    required List<Uint8List> imageBytesList,
    List<int?> imageBytesPerPixel = kBytesPerPixel,
    int width = kWidth,
    int height = kHeight,
    int inputWidth = kInputWidth,
    int inputHeight = kInputHeight,
    List<double> mean = kNormMean,
    List<double> std = kNormStd,
    double minScore = kMinScore,
    int rotation = kRotation,
  }) async {
    final List prediction = await _channel.invokeMethod(
      'predictStreamImage',
      {
        'imageBytesList': imageBytesList,
        'imageBytesPerPixel': imageBytesPerPixel,
        'width': width,
        'height': height,
        'inputWidth': inputWidth,
        'inputHeight': inputHeight,
        'mean': mean,
        'std': std,
        'minScore': minScore,
        'rotation': rotation,
      },
    );

    return prediction;
  }
}
